contracts/oracles/OracleManager.sol

```diff
@@ -105,17 +131,22 @@ contract OracleManager is IOracleManager {
       // Get Previous round data to validate correctness.
       (, , uint256 previousOracleUpdateTimestamp, , ) = chainlinkOracle.getRoundData(oracleRoundIdsToExecute[i] - 1);
 
-      // Check if there was a 'phase change' AND the `_currentOraclePrice` is zero.
-      if ((oracleRoundIdsToExecute[i] >> 64) > (latestExecutedOracleRoundId >> 64) && previousOracleUpdateTimestamp == 0) {
+      // Check if the previous oracle timestamp was zero, but the current one wasn't - then check if there was a phase change.
+      if (previousOracleUpdateTimestamp == 0 && (oracleRoundIdsToExecute[i] >> 64 != (oracleRoundIdsToExecute[i] - 1) >> 64)) {
+        uint16 numberOfPhaseChanges = 1;
+
         // NOTE: if the phase changes, then we want to correct the phase of the update.
         //       There is no guarantee that the phaseID won't increase multiple times in a short period of time (hence the while loop).
         //       But chainlink does promise that it will be sequential.
         // View how phase changes happen here: https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/dev/AggregatorProxy.sol#L335
         while (previousOracleUpdateTimestamp == 0) {
-          // NOTE: re-using this variable to keep gas costs low for this edge case.
-          latestExecutedOracleRoundId = (((latestExecutedOracleRoundId >> 64) + 1) << 64) | uint64(oracleRoundIdsToExecute[i] - 1);
-
-          (, , previousOracleUpdateTimestamp, , ) = chainlinkOracle.getRoundData(latestExecutedOracleRoundId);
+          uint16 prevPhaseId = uint16((oracleRoundIdsToExecute[i] >> 64) - numberOfPhaseChanges++);
+          if (lastRoundId[prevPhaseId] != 0) {
+            // NOTE: re-using this variable to keep gas costs low for this edge case.
+            (, , previousOracleUpdateTimestamp, , ) = chainlinkOracle.getRoundData(lastRoundId[prevPhaseId]);
+          } else {
+            revert NoPhaseIdSet({phaseId: prevPhaseId});
+          }
```

Comment:
We now can use the previous roundIds from the phase change if necessary.

Issue with the previous code is that the round IDs aren't consecutive over phase changes. They reset to to zero each time (in the old code we were assuming they were consecutive - bad internal communication when building out this edge case).

The docs from chainlink are really not great (or in-depth) on the topic - https://docs.chain.link/data-feeds/price-feeds/historical-data - but best to look at the actual contract code on implementation https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/dev/AggregatorProxy.sol#L335